File: 114. Flatten Binary Tree to Linked List/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        if not root:
            return
        
        self.myFlatten(root)
        
    def myFlatten(self, root):
        if not root.right and not root.left:
            return root, root
        elif root.right and not root.left:
            rhead, rtail = self.myFlatten(root.right)
            root.right = rhead
            return root, rtail
        elif not root.right and root.left:
            lhead, ltail = self.myFlatten(root.left)
            root.left = None
            root.right = lhead
            return root, ltail
        else:
            lhead, ltail = self.myFlatten(root.left)
            rhead, rtail = self.myFlatten(root.right)
            root.left = None
            root.right = lhead
            ltail.right = rhead
            return root, rtail

File: 114. Flatten Binary Tree to Linked List/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        if (root == null) {
            return;
        }
        myFlatten(root);
    }
    
    public TreeNode myFlatten(TreeNode root) {
        if (root.left == null && root.right == null) {
            return root;
        } else if (root.left != null && root.right == null) {
            TreeNode tail = myFlatten(root.left);
            root.right = root.left;
            root.left = null;
            return tail;
        } else if (root.left == null && root.right != null) {
            TreeNode tail = myFlatten(root.right);
            return tail;
        } else {
            TreeNode leftTail = myFlatten(root.left);
            TreeNode rightTail = myFlatten(root.right);
            leftTail.right = root.right;
            root.right = root.left;
            root.left = null;
            return rightTail;
        }
    }
}
